mono/tests/test-runner.cs

```diff
@@ -163,20 +163,33 @@ class TestInfo {
 					test_info.Add (new TestInfo { test = s, opt_set = opt });
 			}
 		}		
+		var previous_test_hung = false;
 
 		foreach (TestInfo ti in test_info) {
 			lock (monitor) {
 				while (processes.Count == concurrency) {
-					/* Wait for one process to terminate */
-					Monitor.Wait (monitor);
+					/* Wait for one process to terminate but timeout if it takes too long */
+					bool res = Monitor.Wait (monitor, 1000 * timeout);
+					if (!res) {
+						// Kill and clean up the hung Process
+						if (processes.Count == 1) {
+							previous_test_hung = true;
+  							Process hung = processes[0];
+							ProcessData pd = process_data [hung];
+							pd.CloseStreams ();
+							hung.Kill ();
+							// Wait for the process to finish being killed, but not too long
+							while (!hung.HasExited) {
+								System.Threading.Thread.Sleep(1000);
```

Comment:
this would still loop indefinitely if hung.HasExited is never true. do we need to care about that?